plugin/metadata/metadata.go

```diff
@@ -0,0 +1,40 @@
+package metadata
+
+import (
+	"context"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/miekg/dns"
+)
+
+// Metadater TODO interface needs to be implemented by each plugin willing to provide
+// healthhceck information to the health plugin. Note this method should return
+// quickly, i.e. just checking a boolean status, as it is called every second
+// from the health plugin.
+type Metadater interface {
+	// Metadata returns a TODO
+	Metadata(context.Context, dns.ResponseWriter, *dns.Msg) (context.Context, error)
+}
+
+type Metadata struct {
+	Metadaters []Metadater
+	Next       plugin.Handler
+}
+
+func (m *Metadata) Name() string { return "metadata" }
+
+// ServeDNS implements the plugin.Handler interface.
+func (m *Metadata) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+
+	// Go through all metadaters and collect metadata
+	for _, metadater := range m.Metadaters {
+		if c, err := metadater.Metadata(ctx, w, r); err == nil {
+			ctx = c
+		}
+	}
+
```

Comment:
yes. It is what is expected.
What about optimization ? I mean : should we ensure a more complex mechanism such as we call the metadataer only for a variable that is needed ?

(I am a bit worry by plugin that would provide data aside be useful on some other DNS part ... and we NEVER need to collect the metadata ..)

Reply:
NOTE: was not asked so far. Only a concern raised now that it is written.

Reply:
I tried making a draft version. It seems to be not so straightforward. As I am limited in time - I will implement more simple version for now. I will add it later if I have some time remaining for it.

Reply:
NOTE: I asked JohnB about optimization. And he said it will not work in some cases, like the PolicyPlugin -> because we need the variables to be in the context, but we do not know if the variables will be used ...

So, for now, no change. Thanks